llvm/utils/TableGen/AsmMatcherEmitter.cpp

```diff
@@ -1474,7 +1513,7 @@ void AsmMatcherInfo::buildOperandMatchInfo() {
         OperandMask |= maskTrailingOnes<unsigned>(NumOptionalOps + 1)
                        << (i - NumOptionalOps);
       }
-      if (Op.Class->IsOptional)
+      if (Op.Class->IsOptional && Op.Class->OptionalShouldOffsetCustomParsers)
```

Comment:
I struggle to understand why this is necessary. By the look of it, the flag makes the operand effectively non-optional. Can you give an example demonstrating the problem this flag addresses?

Reply:
Yes I'll take a  shot at explaining.

This is addressing the `ParserMethod` functionality for operands, (eg. FPImmOperand was the one causing me issues)
The way this is done is at operand parsing time (`ARMAsmParser::parseOperand` in this case), `MatchOperandParserImpl` uses the Mnemonic and the number of operands in the OperandVector at that point to check if the operand at that index has a custom ParserMethod.

This causes issues when an operand with a custom ParserMethod comes after an optional operand as the index can change and the optional operand resolution isn't done until match time (after this point). This flag is here to say these optional operands will _always_ be populated populate at parse time so that the indexes of the operands with custom parsers can be known. (See ARMAsmParser.cpp:7062 for where we add the optional operands even if not present in tokens and ARMAsmParser.cpp:7158 after parsing where it is optionally removed).

This is a pretty poor solution getting around the fundamental problem that if we want to have operand classes with custom parsers, where the class information is tied to the instruction definition, then parsing needs to be done at match time where the operand tokens and classes are matched up. But this would require a much deeper change to the parsing and matching machinery.

Reply:
The reason for this flag is because I wanted to be explicit that this is what I was doing and this is behavior I am relying on. The inverse is also possible (always omitting optional operands) and should be as simple as flipping this flag.

Reply:
Heres a little more context for the example: 
Snippet of tableget for VMOVv4f32:
```
def VMOVv4f32 { // InstructionEncoding Instruction InstTemplate Encoding InstARM NeonI NDataI N1ModImm
  field bits<32> Inst = { 1, 1, 1, 1, 0, 0, 1, SIMM{7}, 1, Vd{4}, 0, 0, 0, SIMM{6}, SIMM{5}, SIMM{4}, Vd{3}, Vd{2}, Vd{1}, Vd{0}, 1, 1, 1, 1, 0, 1, 0, 1, SIMM{3}, SIMM{2}, SIMM{1}, SIMM{0} };
  field bits<32> Unpredictable = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
  field bits<32> SoftFail = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
  int Size = 4;
  string DecoderNamespace = "NEONData";
  list<Predicate> Predicates = [HasNEON];
  string DecoderMethod = "DecodeVMOVModImmInstruction";
  bit hasCompleteDecoder = 1;
  string Namespace = "ARM";
  dag OutOperandList = (outs QPR:$Vd);
  dag InOperandList = (ins nImmVMOVF32:$SIMM, pred:$p);
  string AsmString = "vmov${p}.f32  $Vd, $SIMM";
  ...
  ```
With an input like:
```
vmoveq.f32 q2, #0.1
```

In this case before this change the parser for `nImmVMOVF32` (which has parser class `FPImmOperand`) was being applied to `Vd` because `${p}` was present in this case and the parser was assuming that it would not be present (the default behaviour previously). 

This flag and always adding _something_ to the OperandVector was the solution to this. 

Also see `Target.td:957` where I try describe this a bit also.

Reply:
Is `vmoveq.f32 q2, #0.1` supposed to be parsed fine with this patch in place? Here I am at `da3d1607ae9a`, and it doesn't seem to work:

```
$ ./llvm-mc -mcpu=cortex-a8 -triple armv7-apple-darwin -debug <<< 'vmoveq.f32 q2, #0.1'
...
<stdin>:1:1: error: invalid instruction, any one of the following would fix this:
vmoveq.f32 q2, #0.1
^
<stdin>:1:16: note: operand must be a register in range [q0, q15]
vmoveq.f32 q2, #0.1
               ^
<stdin>:1:16: note: invalid operand for instruction
vmoveq.f32 q2, #0.1
               ^
```

The immediate still seems to fail to be parsed for `VMOVv4f32`, if that's the intended instruction:
```
Trying to match opcode VMOVv4f32
  Matching formal operand class MCK_CondCode against actual operand at index 1 (<ARMCC::eq>): match success using generic matcher
  Matching formal operand class MCK__DOT_f32 against actual operand at index 2 ('.f32'): match success using generic matcher
  Matching formal operand class MCK_QPR against actual operand at index 3 (<register q2>): match success using generic matcher
  Matching formal operand class MCK_FPImm against actual operand at index 4 (1036831949): operand match failed, recording near-miss with diag code 0
  Matching formal operand class InvalidMatchClass against actual operand at index 5: actual operand index out of range
```

> The way this is done is at operand parsing time (eg. `ARMAsmParser::parseOperand` in this case), where `MatchOperandParserImpl` uses the Mnemonic and the current OperandVector's size to determine if the operand at that index should use a custom ParserMethod.
> 
> This causes issues when an operand with a custom ParserMethod comes after an optional operand as then the index can change.

So how is this a problem, exactly? The `maskTrailingOnes()` call above is supposed to mark all positions at which an operand with a custom parser may appear, taking into account all optional operands that may precede it. Doesn't it do it correctly in your case for some reason?

> In this case before this change the parser for `nImmVMOVF32` (which has parser class `FPImmOperand`) was being applied to `q2` instead of `#0.1` because `${p}` was present in this case and the parser was assuming that it would not be present (the default behaviour previously).

Again, how you think is this wrong? In presence of optional operands, a custom parser being applied to an umatching operand is normal. Such a custom parser is supposed to return no-match, thus letting know the rest of the parsing infrastructure that this is an operand of some other kind and so other parsers should be tried.

Reply:
Sorry, bad example `vmoveq.f32 q0, #0.125` works with `bin/llvm-mc -debug -triple armv8a-none-eabi -mattr=-fullfp16,+neon -show-encoding <<< "vmov.f32 q0, #0.125"`

Reply:
Hmm thank you for your insight here. It definitely wasn't working without this, but perhaps that was a bug in a different area and not what I thought it was. 

I think maybe this bug stems from `ARMAsmParser::parseFPImm` in several situations produces an error rather than `NoMatch`. I will see if I can change this to get around this issue a different way.

Reply:
It turns out this was a symptom of the remaining hacks surrounding VPTPred not being optional ):
I am going to try make this optional also, thank you for catching this

Reply:
@kosarev I have resolved this. I had misdiagnosed an issue where certain custom parsers where behaving poorly when being applied to non-matching operands. The new patch also makes VPT operand optional and fixes these cases, removing the need for this option.